crates/solidity/testing/sanctuary/src/tests.rs

```diff
@@ -100,21 +104,45 @@ pub fn run_test(file: &SourceFile, events: &Events) -> Result<()> {
 
     let parser = Parser::create(version.clone())?;
     let output = parser.parse(NonterminalKind::SourceUnit, &source);
+    let source_id = file.path.strip_repo_root()?.unwrap_str();
 
     if output.is_valid() {
-        events.test(TestOutcome::Passed);
-        return Ok(());
-    }
+        let mut outcome = TestOutcome::Passed;
 
-    events.test(TestOutcome::Failed);
+        let bindings = create_bindings(&version, source_id, &output)?;
 
-    let with_color = true;
-    let source_id = file.path.strip_repo_root()?.unwrap_str();
+        for reference in bindings.all_references() {
+            if reference.get_file().is_system() {
+                // skip built-ins
+                continue;
+            }
+            // We're not interested in the exact definition a reference resolves
+            // to, so we lookup all of them and fail if we find none.
+            if reference.definitions().is_empty() {
+                outcome = TestOutcome::Failed;
+                let cursor = reference.get_cursor().unwrap();
+                let report = format!(
```

Comment:
nit: WDYT of using `diagnostic::render()` below to create the `report`? helps with visually debugging multiple errors..
Not blocking though..

Reply:
Oh, that's a good idea. For some reason I thought `diagnostic::render()` was tied to parsing errors, but I see it's more generic. I'll change it.